### PR TITLE
[RELENG-4725] Add github sso autentication to nginx

### PR DESCRIPTION
**Pull request by erwan-b**

### CODE REVIEW

File: lua/github_access.lua [hunk 1]
```diff
@@ -0,0 +1,63 @@
+local b64 = require "ngx.base64"
+
+
+function wrong_credentials()
+   ngx.header.content_type = 'text/plain'
+   ngx.header['WWW-Authenticate'] = 'Basic realm="Access to the Scality Artifacts", charset="UTF-8"'
+   ngx.status = ngx.HTTP_UNAUTHORIZED
+   ngx.log(ngx.STDERR, '401 Access Denied')
+   return ngx.exit(ngx.HTTP_OK)
```
Comment (bertranddemiddelaer) [inline lua/github_access.lua:9]: HTTP_OK ? Did not you want HTTP_UNAUTHORIZED instead ?

File: lua/github_access.lua [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+end
+
+
+function not_allowed()
+   ngx.header.content_type = 'text/plain'
+   ngx.status = ngx.HTTP_FORBIDDEN
+   ngx.log(ngx.STDERR, '403 Forbidden response')
+   return ngx.exit(ngx.HTTP_OK)
```
Comment (bertranddemiddelaer) [inline lua/github_access.lua:17]: Did you mean HTTP_FORBIDDEN ?

File: lua/github_access.lua [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+end
+
+function authenticate(auth)
+    divider = auth:find(':')
+    local username = auth:sub(0, divider-1)
+    local password = auth:sub(divider+1)
+    local resp = {}
+
+    local res =  ngx.location.capture("/force_external_request/orgs/scality/members/" .. username)
```
Comment (bertranddemiddelaer) [inline lua/github_access.lua:26]: s/external/github/ no ?

File: lua/github_access.lua [hunk 4]
```diff
@@ -0,0 +1,63 @@
+local b64 = require "ngx.base64"
```
Comment (bertranddemiddelaer) [inline lua/github_access.lua:1]: Not sure it is needed. Seems like dead code.

File: lua/github_access.lua [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+   ngx.header.content_type = 'text/plain'
+   ngx.status = ngx.HTTP_FORBIDDEN
+   ngx.log(ngx.STDERR, '403 Forbidden response')
+   return ngx.exit(ngx.HTTP_FORBIDDEN)
+end
+
+function authenticate(auth)
+    divider = auth:find(':')
+    local username = auth:sub(0, divider-1)
+    local password = auth:sub(divider+1)
```
Comment (bertranddemiddelaer) [inline lua/github_access.lua:20]: This variable is not used anywhere ?

**Review by bertranddemiddelaer — COMMENTED** [review]
This PR needs to change tests to prove that auth is working. Good news is that right now, unmodified tests show that the client that does not provide any credentials sees its access refused.

File: lua/github_access.lua [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+   ngx.log(ngx.STDERR, '403 Forbidden response')
+   return ngx.exit(ngx.HTTP_FORBIDDEN)
+end
+
+function authenticate(auth)
+    divider = auth:find(':')
+    local username = auth:sub(0, divider-1)
+    local resp = {}
+
+    local res =  ngx.location.capture("/force_github_request/orgs/scality/members/" .. username)
```
Comment (tcarmet) [inline lua/github_access.lua:22]: scality org should not be hardcoded and should rather be a variable.

**Review by tcarmet — CHANGES_REQUESTED** [review]
Overall the lua code and what we're doing here looks more than good to me.

However I would add a couple of requirements.
* I should be able to deactivate or activate this setting. It should not be mandatory. I want to be able to run artifacts on my local machine without requiring it to set it up with github auth.
* Let's not hardcode the organization we are checking in the code, and set a proper variable to set this up.

On a side note: Do not forget to update the helm chart accordingly with the new settings if required

**Review by tcarmet — CHANGES_REQUESTED** [review]
Thanks for applying my comment! It's looking good.

The last thing I see would be to update the helm chart in order to be able to set the new variables in Artifacts' runtime.
Check `helm/charts/artifacts`

**Comment by erwan-b** [comment]
`It seems that authentication from client point of view is being done via a header, where today in artifacts it's being done via basic auth. Can you confirm to me that we are still able to login via basic auth? Some clients like apt/yum will require it. @bertranddemiddelaer I'll let you check that with @erwan-b (After re reading the code I think it's already implemented, so I'm probably mistaken here, but could you please double check heart )
`

Basic auth is done by header :+1:

File: charts/artifacts/templates/secrets.yaml [hunk 7]
```diff
@@ -14,3 +14,6 @@ data:
   AWS_SECRET_ACCESS_KEY: {{ .Values.cloudStorage.awssecretaccesskey | b64enc | quote }}
   AWS_BUCKET_PREFIX: {{ .Values.cloudStorage.awsbucketprefix | b64enc | quote }}
   ENDPOINT_URL: {{ .Values.cloudStorage.endpointUrl | b64enc | quote }}
+  GITHUB_API_URL: {{ .Values.cloudStorage.githubApiUrl | b64enc | quote }}
+  GITHUB_API_COMPANY: {{ .Values.cloudStorage.githubApiCompany | b64enc | quote }}
+  GITHUB_API_ENABLED: {{ .Values.cloudStorage.githubApiEnabled | b64enc | quote }}
```
Comment (tcarmet) [inline charts/artifacts/templates/secrets.yaml:19]:
```suggestion
  GITHUB_API_URL: {{ .Values.githubApiUrl | b64enc | quote }}
  GITHUB_API_COMPANY: {{ .Values.githubApiCompany | b64enc | quote }}
  GITHUB_API_ENABLED: {{ .Values.githubApiEnabled | b64enc | quote }}
```

File: charts/artifacts/values.yaml [hunk 8]
```diff
@@ -24,6 +24,18 @@ cloudStorage:
...
+  githubApiUrl: 'https://api.github.com'
+
+  ## githubApiCompany The github company to check the user is in
+  ##
+  githubApiCompany: ''
+
+  ## githubApiEnabled set the github basic auth check enabled or not
+  ##
+  githubApiEnabled: 'true'
+
```
Comment (tcarmet) [inline charts/artifacts/values.yaml:38]:
```suggestion
## githubApiUrl contains the api url to check the auth connection
##
githubApiUrl: 'https://api.github.com'

## githubApiCompany The github company to check the user is in
##
githubApiCompany: ''

## githubApiEnabled set the github basic auth check enabled or not
##
githubApiEnabled: 'true'
```

**Review by tcarmet — COMMENTED** [review]
Last comment, those variables do not belong in the cloudstorage section. Cloudstorage is meant for communication with S3 buckets.

**Review by tcarmet — APPROVED** [review]

**Pull request merged**